### PR TITLE
Support mounting existing FSx for Ontap and OpenZFS

### PR DESCRIPTION
**Pull request by hanwen-pcluste**
1. Add section of FsxOntap and FsxOpenZfs settings in configuration file:
```
SharedStorage:
  - MountDir: /my/mount/point1
    Name: name1
    StorageType: FsxOpenZfs
    FsxOpenZfsSettings:
      VolumeId: fsvol-12345678901234567
  - MountDir: /my/mount/point2
    Name: name2
    StorageType: FsxOntap
    FsxOntapSettings:
      VolumeId: fsvol-12345678901234567
```
2. Use class inheritance in cluster_config.py for FSx for Lustre, Optap, OpenZFS
3. Pass two more attributes to cookbook:
`fsx_volume_junction_paths`: It is a comma separated list. There are strings between the comma if it is FSx for Ontap and OpenZFS. There is empty if it is FSx for Lustre.
`fsx_fs_types`: It is a comma separated list. It tells the cookbook what mount method to use according to the file system (Lustre, Ontap, OpenZFS)
4. Improve FSx security group validator to check different ports for different file system
5. Improve integration test to test the new file system. Move fsx_factory to conftest, so it can be reused by other tests

Test:
Manual tests are done with test_ad_integration and test_cluster_in_no_internet_subnet

Signed-off-by: Hanwen <hanwenli@amazon.com>

### Tests
Improved test_multiple_fsx tests to check Ontap and OpenZFS,
Manual tests are done with test_ad_integration and test_cluster_in_no_internet_subnet

### References
* This PR works together with https://github.com/aws/aws-parallelcluster-cookbook/pull/1446

### Checklist
- [ ] Make sure you are pointing to **the right branch** and add a label in the PR title (i.e. **2.x** vs **3.x**)
- [ ] Check all commits' messages are clear, describing what and why vs how.
- [ ] Make sure **to have added unit tests or integration tests** to cover the new/modified code.
- [ ] Check if documentation is impacted by this change.

Please review the [guidelines for contributing](../CONTRIBUTING.md) and [Pull Request Instructions](https://github.com/aws/aws-parallelcluster/wiki/Git-Pull-Request-Instructions).

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: cli/src/pcluster/aws/aws_resources.py
Comment (gmarciani) [inline cli/src/pcluster/aws/aws_resources.py]: since `FSX_ONTAP` is missing, I assume there will be some differences in the way we interact with Ontap. May you please clarify?

File: cli/src/pcluster/aws/aws_resources.py
Comment (gmarciani) [inline cli/src/pcluster/aws/aws_resources.py]: Why `MountName` is empty for FsxOpenZsf and FsxOntap? In the config file we are able to set the mount name for every type of FS so this looks like unexpected. May you please clarify both in this conversation and with a comment on code?

File: cli/src/pcluster/aws/aws_resources.py
Comment (gmarciani) [inline cli/src/pcluster/aws/aws_resources.py]: Same as above: I'm not getting why the `DnsName` is returned only for FsxLustre and FsxOpenZfs and it's not for FsxOntap. May you please clarify both in this conversation and with a comment on code?

File: cli/src/pcluster/aws/fsx.py
Comment (gmarciani) [inline cli/src/pcluster/aws/fsx.py:21]: I suppose that *svm* stands for storage virtual machines, but I needed to go through the rest of the code to understand it? I suggest to avoid acronyms whenever it is possible.

File: cli/src/pcluster/aws/fsx.py
Comment (gmarciani) [inline cli/src/pcluster/aws/fsx.py:56]: We are caching without cache expiration strategy. 
I suppose that if we are caching we are expecting this command to be executed a lot of times.
Are we sure that the cached data is not invalidated over time?

File: cli/src/pcluster/aws/fsx.py
Comment (gmarciani) [inline cli/src/pcluster/aws/fsx.py:60]: Here we are introducing a new AWS API call. Let's be sure that we are aligning the user policies accordingly in our public docs.
If it is not yet included, we must document it and review it with the security team.

File: cli/src/pcluster/aws/fsx.py
Comment (gmarciani) [inline cli/src/pcluster/aws/fsx.py:80]: Same as above: let's be sure that this call is allowed by the user policy in our public docs.
If it is not yet included, we must document it and review it with the security team.

File: cli/src/pcluster/config/cluster_config.py
Comment (gmarciani) [inline cli/src/pcluster/config/cluster_config.py:316]: Here we are assuming that there cannot be the case of missing both file system data and file system id. Are we sure that our validators already prevent this?

Comment (gmarciani) [inline cli/src/pcluster/config/cluster_config.py:316]: I don't see the class attribute `file_system_id` defined anywhere in the class constructor. Where are we expecting it to be defined? If we are expecting this class attribute to be defined in a subclass, I think it's not a good practice: a class should reference its own data. In this specific case, I think we should move `file_system_id` in the base class, right?

File: cli/src/pcluster/config/cluster_config.py
Comment (gmarciani) [inline cli/src/pcluster/config/cluster_config.py:352]: Isn't `fsx_storage_type` implicit in `file_system_type`? What about reducing knowledge duplication?

File: cli/src/pcluster/config/cluster_config.py
Comment (gmarciani) [inline cli/src/pcluster/config/cluster_config.py:431]: According to PR description, it seems that both Ontap and OpenZFS have junction paths. Why this concept is missing in`ExistingFsxOpenZfs ` and it is implement as a property in the below `ExistingFsxOntap`?

Comment (gmarciani) [reply cli/src/pcluster/config/cluster_config.py:431]: According to the code, it seems that the two file systems share the concept of volumes, but not the concept of junction paths, so why are we using the naming *junction path instead of *volumes* is cookbook attributes?

File: cli/src/pcluster/config/cluster_config.py
Comment (gmarciani) [inline cli/src/pcluster/config/cluster_config.py]: Why are we ignoring this flake rule?

File: cli/src/pcluster/config/cluster_config.py
Comment (gmarciani) [inline cli/src/pcluster/config/cluster_config.py]: Code duplication: these two lines are equivalent to the above ones.

File: cli/src/pcluster/config/cluster_config.py
Comment (gmarciani) [inline cli/src/pcluster/config/cluster_config.py:1562]: Are we sure that the batch validator must be applied only when FsxLustre is used?

File: cli/src/pcluster/constants.py
Comment (gmarciani) [inline cli/src/pcluster/constants.py:66]: May you please add a comment with links to the official documentation that declares these ports, if any?

File: cli/src/pcluster/schemas/cluster_schema.py
Comment (gmarciani) [inline cli/src/pcluster/schemas/cluster_schema.py]: 1. We are duplicating the same volume regexp for OpenZFS and Ontap. What about putting it as a constant in constants.py file close to where we define file system related constants?
2. May you please link the official doc where the regexp is documented?
3. Here we are including in the schema only the parameter `volume_id`, but in other parts of this PR we are assuming that the settings may have other parameters, e.g. StorageVirtualMachineId, VolumeJunctionPath. So, why are they excluded from the schema?

File: cli/src/pcluster/schemas/cluster_schema.py
Comment (gmarciani) [inline cli/src/pcluster/schemas/cluster_schema.py]: The following strings are duplicated many times: "FsxLustre", "FsxOpenZfs", "FsxOntap".
What about defining them as constants instead or re-writing them many times?

File: cli/src/pcluster/schemas/cluster_schema.py
Comment (gmarciani) [inline cli/src/pcluster/schemas/cluster_schema.py]: I'm not getting why in the lines from the previous diff block `storage_type` is expected to have values "FsxLustre", "FsxOpenZfs", "FsxOntap"; whereas in this diff block it is expected to have "fsx_lustre", "fsx_open_zfs" and "fsx_ontap". I think we should consider to align these naming conventions to reduce the risk of errors.

File: cli/src/pcluster/schemas/cluster_schema.py
Comment (gmarciani) [inline cli/src/pcluster/schemas/cluster_schema.py:542]: the missing final else within the first if makes `storage_type` potentially uninitialized

File: cli/src/pcluster/templates/cluster_stack.py
Comment (gmarciani) [inline cli/src/pcluster/templates/cluster_stack.py:616]: Overall in the PRs we mention volumes, junctions paths, volume path and volume junction paths. We need to be more clear about their different roles and relations.

File: cli/src/pcluster/validators/cluster_validators.py
Comment (gmarciani) [inline cli/src/pcluster/validators/cluster_validators.py:414]: What is the upper bound of the number of calls we are expecting and how much are we reducing them?
Are we sure that aiming to this reduction does not make the code more error prone or, even worst, make the validation less clear from the customer perspective?

File: cli/tests/pcluster/validators/test_cluster_validators.py
Comment (gmarciani) [inline cli/tests/pcluster/validators/test_cluster_validators.py:548]: Is the `r` prefix really necessary?
Furthermore, here we are assuming ports 988, so basically we are covering only the case of FsxLustre. What about the other ones?

File: tests/integration-tests/tests/storage/test_fsx_lustre.py
Comment (gmarciani) [inline tests/integration-tests/tests/storage/test_fsx_lustre.py:277]: Aren't we assuming to use TCP? Why are we including also UDP?

File: cli/src/pcluster/config/cluster_config.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py:431]: FSx names the path differently, e.g. `junction path` in Ontap vs `volume path` in OpenZFS. But the concept is the same. Therefore, the property names are different in the code. However, when passing the path to cookbook, I unified them into `fsx_volume_junction_paths`

If we pass volumes instead of paths to cookbook, an additional AWS API is required in Cookbook. AWS API call should be done on the CLI to reduce API throttling when scaling up, and to better support cluster in isolated subnet.

File: cli/src/pcluster/config/cluster_config.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py:352]: Probably `fsx_storage_type` should be renamed as `fsx_lustre_storage_type`?

File: cli/src/pcluster/aws/aws_resources.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/aws/aws_resources.py]: Mount name is specific to FSx Lustre, and is retrieved from AWS API

In config file, we can specify mount dir for all the file systems.

File: cli/src/pcluster/aws/fsx.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/aws/fsx.py:21]: SVM stands for storage virtual machine

File: cli/src/pcluster/aws/fsx.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/aws/fsx.py:56]: Within a single pcluster command execution, we can assume the API response is the same. Across multiple pcluster commands. the cache is always cleared

File: cli/src/pcluster/config/cluster_config.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py:316]: `file_system_data` is only used when existing FSx is provided, when `file_system_id` always exists.
If a new FSx Lustre is being created, the code does not use `file_system_data`

Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py:316]: The retrieval of `file_system_id` is different depending on the file system type (Lustre, Ontap, OpenZFS)

File: cli/src/pcluster/config/cluster_config.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py]: It is function too complex warning. An alternative is to split the function into different function. What do you think?

File: cli/src/pcluster/config/cluster_config.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py]: Done

File: cli/src/pcluster/config/cluster_config.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py:1562]: Good catch. I fix this along with other bugs from the renaming

File: cli/src/pcluster/constants.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/constants.py:66]: Done

File: cli/src/pcluster/schemas/cluster_schema.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: 1. We only duplicated once here. Can we keep them here?
2. Other regex in this file is not commented. The document is here https://docs.aws.amazon.com/fsx/latest/APIReference/API_DescribeVolumes.html#FSx-DescribeVolumes-request-VolumeIds. Should we add it to the doc? Are we assuming the link is valid permenantly?
3. Other information is implied by volume id

File: cli/src/pcluster/schemas/cluster_schema.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: Done

File: cli/src/pcluster/schemas/cluster_schema.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: This is some magic with the schema. It is not easy to align them

File: cli/src/pcluster/validators/cluster_validators.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/validators/cluster_validators.py:414]: The upper bound is 50 file systems. We are reducing them to constant number of calls regardless of the number of the file system

From the customer perspective, nothing is changed

File: cli/tests/pcluster/validators/test_cluster_validators.py
Comment (hanwen-pcluste) [reply cli/tests/pcluster/validators/test_cluster_validators.py:548]: `r` is needed when writing regex with escapes.
Yes , here we only cover Lustre

File: tests/integration-tests/tests/storage/test_fsx_lustre.py
Comment (hanwen-pcluste) [reply tests/integration-tests/tests/storage/test_fsx_lustre.py:277]: Ontap and OpenZFS requires both

File: cli/src/pcluster/aws/aws_resources.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/aws/aws_resources.py]: Because the condition statements below only affects some of the file system types

File: cli/src/pcluster/aws/aws_resources.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/aws/aws_resources.py]: Done

File: cli/src/pcluster/aws/aws_resources.py
Comment (gmarciani) [reply cli/src/pcluster/aws/aws_resources.py]: Sure, the question is: why Ontap differs w.r.t. the others.

File: cli/src/pcluster/aws/aws_resources.py
Comment (gmarciani) [reply cli/src/pcluster/aws/aws_resources.py]: Thanks for adding a comment.
You say "Ontap has DNS name on storage virtual machines.".
So, we are not retrieving DNS name for Ontap here because in terms of clean coding is not correct to retrieve it from here, or because we do not need to retrieve it at all?

File: cli/src/pcluster/aws/fsx.py
Comment (gmarciani) [reply cli/src/pcluster/aws/fsx.py:56]: Got it. So, within a single pcluster command how many times are we expecting to execute such API calls to justify a caching strategy?

File: cli/src/pcluster/config/cluster_config.py
Comment (gmarciani) [reply cli/src/pcluster/config/cluster_config.py:316]: So, file_system_id is an "abstract" attribute whose value is set within file system specific sub-classes.
Such abstract behaviour should be implemented in a more robust way, e.g. defined as an `@abstractmethod @property` within the super-class. See https://towardsdatascience.com/how-to-use-abstract-classes-in-python-d4d2ddc02e90

File: cli/src/pcluster/config/cluster_config.py
Comment (gmarciani) [reply cli/src/pcluster/config/cluster_config.py:352]: Checked the [docs](https://docs.aws.amazon.com/parallelcluster/latest/ug/SharedStorage-v3.html#yaml-SharedStorage-FsxLustreSettings-StorageType). there is no knowledge duplication here:`fsx_storage_type` (HDD,SSD) cannot be retrieved from `file_system_type` (LUSTRE).

Let's keep it as it is.

File: cli/src/pcluster/config/cluster_config.py
Comment (gmarciani) [reply cli/src/pcluster/config/cluster_config.py:431]: For the cookbook: let's keep concepts separated and self-contained. Generally speaking, variables should be associated to a single precise concept. We should avoid to have a variable like `fsx_volume_junction_paths` if it can contain volume ids OR paths. We should have a variable for junction paths and a variable for volume ids

File: cli/src/pcluster/config/cluster_config.py
Comment (gmarciani) [reply cli/src/pcluster/config/cluster_config.py]: I think it would be better to split as long as it improves readability and maintainability, that's generally the case.

File: cli/src/pcluster/schemas/cluster_schema.py
Comment (gmarciani) [reply cli/src/pcluster/schemas/cluster_schema.py]: 1. I think it's better to not duplicating the information at all
2. Thanks for sharing the link. I think adding the link in a comment would add some value because even if invalidated sooner or later it will provide information about what we considered the source of thruth for that regexp, i.e. the FSX DescribeVolumes API.
3. That does not match with what I understood from the design discussions. Let's have a follow up on this, please.

File: cli/src/pcluster/schemas/cluster_schema.py
Comment (gmarciani) [reply cli/src/pcluster/schemas/cluster_schema.py]: If something appears as magic, then we should improve something I suppose :-) 
Let's be sure to track this in a refactoring task.

File: cli/src/pcluster/schemas/cluster_schema.py
Comment (gmarciani) [reply cli/src/pcluster/schemas/cluster_schema.py:542]: Still not seeing the fix. By the way, what about using a dictionary instead of a nested if?

File: cli/tests/pcluster/validators/test_cluster_validators.py
Comment (gmarciani) [reply cli/tests/pcluster/validators/test_cluster_validators.py:548]: Shouldn't we extend  it to cover also the other file systems?

File: cli/src/pcluster/aws/aws_resources.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/aws/aws_resources.py]: The DNS name is needed. But for clean coding, we need to get the DNS name form storage virtual machine

File: cli/src/pcluster/aws/fsx.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/aws/fsx.py:56]: The maximum file system number is 50, So we are comparing 1 call vs 50 calls

File: cli/src/pcluster/config/cluster_config.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py:431]: Volume id is not passed to cookbook. All the values in fsx_volume_junction_paths are paths

File: cli/src/pcluster/schemas/cluster_schema.py
Comment (gmarciani) [reply cli/src/pcluster/schemas/cluster_schema.py]: On point 3: checked the final decision in design and we need only volume id in the schema. That's fine.

File: cli/src/pcluster/config/cluster_config.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py:316]: For existing FSx for Lustre, `file_system_id` is a field from cluster configuration.
For existing FSx for Ontap and OpenZFS, `file_system_id` is property, retrieved using volume id.

So it not easy to do anything in the superclass

File: cli/src/pcluster/config/cluster_config.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py]: Done

File: cli/src/pcluster/schemas/cluster_schema.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py]: Done

File: cli/src/pcluster/schemas/cluster_schema.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/schemas/cluster_schema.py:542]: Done

File: cli/tests/pcluster/validators/test_cluster_validators.py
Comment (hanwen-pcluste) [reply cli/tests/pcluster/validators/test_cluster_validators.py:548]: Done

File: cli/src/pcluster/aws/aws_resources.py
Comment (gmarciani) [reply cli/src/pcluster/aws/aws_resources.py]: We had a offline follow up on this. That's fine.

File: cli/src/pcluster/aws/fsx.py
Comment (gmarciani) [reply cli/src/pcluster/aws/fsx.py:56]: Ok, so please correct me if I'm wrong.

We are caching SVM data because we can mount different volumes having the same SVM, so with caching we are limiting the number of times we retrieve SVM data.

Correct?

File: cli/src/pcluster/config/cluster_config.py
Comment (gmarciani) [inline cli/src/pcluster/config/cluster_config.py:311]: Is it possible to test the inclusion of SharedStorageNameValidator (see codecov comment below)?

File: cli/src/pcluster/config/cluster_config.py
Comment (gmarciani) [reply cli/src/pcluster/config/cluster_config.py:316]: I'm not sure about that. Btw, let's consider this as as non blocking comment.

File: cli/src/pcluster/config/cluster_config.py
Comment (gmarciani) [reply cli/src/pcluster/config/cluster_config.py:431]: Got it, thanks

File: cli/src/pcluster/config/cluster_config.py
Comment (gmarciani) [reply cli/src/pcluster/config/cluster_config.py:1554]: Did you fix it?

File: cli/src/pcluster/schemas/cluster_schema.py
Comment (gmarciani) [reply cli/src/pcluster/schemas/cluster_schema.py:542]: Thanks

File: cli/tests/pcluster/validators/test_cluster_validators.py
Comment (gmarciani) [reply cli/tests/pcluster/validators/test_cluster_validators.py:548]: Thanks

File: tests/integration-tests/conftest.py
Comment (gmarciani) [inline tests/integration-tests/conftest.py]: What about the Code Scanning comment below?

File: cli/src/pcluster/aws/aws_resources.py
Comment (gmarciani) [reply cli/src/pcluster/aws/aws_resources.py]: Questions:
1. What is the expected content of `file_system_data.get("DNSName")` for ONTAP?
1. What about adding the DNSName also for ONTAP within `file_system_data` before passing it to the constructor of `FsxFileSystemInfo`?

Comment (hanwen-pcluste) [reply cli/src/pcluster/aws/aws_resources.py]: 1. There is no such key in the response. So the result is None
2. the DNSName of Ontap is on the storage virtual machine. Why adding it to file system data?

File: cli/src/pcluster/aws/fsx.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/aws/fsx.py:56]: Yes.
It is not only for the same SVM, but also for different SVMs.
There could've been a boto3 call to describe a list of SVMs, or many single calls to describe each SVM.
Here, we are using the first approach to reduce the number of calls to once

File: cli/src/pcluster/config/cluster_config.py
Comment (hanwen-pcluste) [reply cli/src/pcluster/config/cluster_config.py:311]: Done

File: tests/integration-tests/conftest.py
Comment (hanwen-pcluste) [reply tests/integration-tests/conftest.py]: Done

File: cli/src/pcluster/config/cluster_config.py
Comment (gmarciani) [reply cli/src/pcluster/config/cluster_config.py:316]: Not sure about that, but let's consider this as not blocking for the moment.

File: cli/src/pcluster/templates/cluster_stack.py
Comment (gmarciani) [inline cli/src/pcluster/templates/cluster_stack.py]: Not a blocker, just an improvement: what about setting the `dns_name` value within `shared_fsx.existing_dns_name`, like in the other cases (probably with different Boto calls)?

Comment (hanwen-pcluste) [reply cli/src/pcluster/templates/cluster_stack.py]: Done

**Review by gmarciani — APPROVED** [review]

**Pull request merged**
